storagenode/vouchers/service.go

```diff
@@ -7,16 +7,165 @@ import (
 	"context"
 	"time"
 
+	"github.com/zeebo/errs"
+	"go.uber.org/zap"
+	"golang.org/x/sync/errgroup"
+	monkit "gopkg.in/spacemonkeygo/monkit.v2"
+
+	"storj.io/storj/internal/sync2"
+	"storj.io/storj/pkg/kademlia"
 	"storj.io/storj/pkg/pb"
 	"storj.io/storj/pkg/storj"
+	"storj.io/storj/pkg/transport"
+	"storj.io/storj/storagenode/trust"
+)
+
+var (
+	// VoucherError represents errors with vouchers
+	VoucherError = errs.Class("voucher")
+
+	mon = monkit.Package()
 )
 
 // DB implements storing and retrieving vouchers
 type DB interface {
 	// Put inserts or updates a voucher from a satellite
 	Put(context.Context, *pb.Voucher) error
-	// GetExpiring retrieves all vouchers that are expired or about to expire
-	GetExpiring(context.Context, time.Duration) ([]storj.NodeID, error)
 	// GetValid returns one valid voucher from the list of approved satellites
 	GetValid(context.Context, []storj.NodeID) (*pb.Voucher, error)
+	// NeedVoucher returns true if a voucher from a particular satellite is expired, about to expire, or does not exist
+	NeedVoucher(context.Context, storj.NodeID, time.Duration) (bool, error)
+}
+
+// Config defines configuration for requesting vouchers.
+type Config struct {
+	Interval         int `help:"number of days between voucher service iterations" default:"7"`
+	ExpirationBuffer int `help:"buffer period of X days into the future. If a voucher would expire within this period, send a request for renewal" default:"7"`
```

Comment:
we might want to hard code `ExpirationBuffer` to be some value longer than `Interval`

Reply:
also can this be a time.Duration instead of an int?

Reply:
Jessica and I were thinking it might be better to hard-code this to be == `Interval` or `Interval` + (some extra duration). What do you think?